### PR TITLE
Updated SignalServiceKit and SignalProtocolKit.

### PR DESCRIPTION
**Pull request by Elland**
Please consider doing some live testing messaging other devices.

### CODE REVIEW

File: Token.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1317,6 +1321,7 @@
 				9FB36EA81E7C27A300DB6F86 /* menuButton.m4a in Resources */,
 				149F9AD21E72E29A00FB74AA /* SF-UI-Text-Medium.otf in Resources */,
 				149F9AD31E72E29A00FB74AA /* SF-UI-Text-Bold.otf in Resources */,
+				9F3FF7F61EB3341F002D0293 /* uk.svg in Resources */,
```
Comment (yuliaveres) [inline Token.xcodeproj/project.pbxproj]: what is that image for ? :)

**Comment by 3lvis** [comment]
Looks fine to me, gonna do some testing.

File: Token/Controllers/Messaging/MessagesViewController.swift [hunk 2]
```diff
@@ -699,7 +711,7 @@ extension MessagesViewController: ImagePickerDelegate {
                 let timestamp = NSDate.ows_millisecondsSince1970(for: Date())
                 let outgoingMessage = TSOutgoingMessage(timestamp: timestamp, in: self.thread, messageBody: "")
 
-                self.messageSender.sendAttachmentData(imageData, contentType: "image/jpeg", filename: "", in: outgoingMessage, success: {
+                self.messageSender.sendAttachmentData(imageData, contentType: "image/jpeg", filename: "image.jpeg", in: outgoingMessage, success: {
```
Comment (yuliaveres) [inline Token/Controllers/Messaging/MessagesViewController.swift]: the image name not the best one :/ but dnt have better idea now

File: Token/Models/Message.swift [hunk 3]
```diff
@@ -32,8 +32,15 @@ public class Message: NSObject, NOCChatItem {
                 let attachment = TSAttachment.fetch(withUniqueID: attachmentId)!
                 if attachment is TSAttachmentPointer {
                     return [#imageLiteral(resourceName: "placeholder")]
-                } else if let stream = attachment as? TSAttachmentStream, let image = stream.image() {
-                    return [image]
+                } else if let stream = attachment as? TSAttachmentStream {
+                    if stream.isVideo(), let thumbnail = stream.image() {
+                        return [thumbnail]
```
Comment (yuliaveres) [inline Token/Models/Message.swift]: are we okay with multiple returns in a method ? 😢

File: Token/Models/Message.swift [hunk 4]
```diff
@@ -32,8 +32,15 @@ public class Message: NSObject, NOCChatItem {
...
+                    } else if stream.isImage(), let image = stream.image() {
```
Comment (yuliaveres) [inline Token/Models/Message.swift]: also, both checks look similar, only namings are different, can we use common logic for video / image stream's image() ?

File: Token/Models/Message.swift [hunk 5]
```diff
@@ -32,8 +32,15 @@ public class Message: NSObject, NOCChatItem {
...
+                        return [image]
+                    } else if let url = stream.mediaURL() {
+                        print(url)
```
Comment (yuliaveres) [inline Token/Models/Message.swift]: do we need print here ? if not, should remember to remove it 😉

File: Carthage/Build/iOS/ImagePicker.framework/Headers/ImagePicker-Swift.h [hunk 6]
```diff
@@ -198,6 +198,7 @@ SWIFT_CLASS("_TtC11ImagePicker21ImagePickerController")
 @property (nonatomic) NSInteger imageLimit;
 @property (nonatomic) BOOL startOnFrontCamera;
 @property (nonatomic, copy) NSString * _Nullable doneButtonTitle;
+- (nonnull instancetype)initWithNibName:(NSString * _Nullable)nibNameOrNil bundle:(NSBundle * _Nullable)nibBundleOrNil OBJC_DESIGNATED_INITIALIZER;
```
Comment (yuliaveres) [inline Carthage/Build/iOS/ImagePicker.framework/Headers/ImagePicker-Swift.h]: we should keep consistency between nullability specifiers nullable / _Nullable
(as for me nullable looks more readable.)

Comment (Elland) [inline Carthage/Build/iOS/ImagePicker.framework/Headers/ImagePicker-Swift.h]: That's not our code tho. 👍

**Comment by 3lvis** [comment]
On this update, the "sending" indicator keeps spinning meanwhile in `master` it works fine.

image uploaded from ios 2

File: Token/Models/Message.swift (hunk 4 above)
Comment (Elland) [inline Token/Models/Message.swift]: They're only the same for now, because we didn't get the design yet. 👍

File: Token/Models/Message.swift (hunk 3 above)
Comment (Elland) [inline Token/Models/Message.swift]: Instead of?

File: Token/Controllers/Messaging/MessagesViewController.swift (hunk 2 above)
Comment (Elland) [inline Token/Controllers/Messaging/MessagesViewController.swift]: We could probably extract something from the attachment, once we get there.

File: Token.xcodeproj/project.pbxproj (hunk 1 above)
Comment (Elland) [inline Token.xcodeproj/project.pbxproj]: Testing. Gonna delete it!

File: Token/Models/Message.swift (hunk 3 above)
Comment (3lvis) [inline Token/Models/Message.swift]: Something like this:

```swift
public var images: [UIImage] {  
    var result = [UIImage]

    if self.signalMessage.hasAttachments() {
        if let attachmentId = (signalMessage.attachmentIds as? [String])?.first {
            let attachment = TSAttachment.fetch(withUniqueID: attachmentId)!
            if attachment is TSAttachmentPointer {
                result = [#imageLiteral(resourceName: "placeholder")]
            } else if let stream = attachment as? TSAttachmentStream {
                if stream.isVideo(), let thumbnail = stream.image() {
                    result = [thumbnail]
                } else if stream.isImage(), let image = stream.image() {
                    result = [image]
                } else if let url = stream.mediaURL() {
                    result = [#imageLiteral(resourceName: "placeholder")]
                }
            }
        }
    }

    return result
}
```

File: Token/Models/Message.swift (hunk 4 above)
Comment (yuliaveres) [inline Token/Models/Message.swift]: ok

File: Token/Models/Message.swift (hunk 3 above)
Comment (Elland) [inline Token/Models/Message.swift]: Sure, but why? 💭

File: Token/Models/Message.swift [hunk 7]
```diff
@@ -127,5 +127,17 @@ public class Message: NSObject, NOCChatItem {
         self.isActionable = shouldProcess && !isOutgoing && (sofaWrapper?.type == .paymentRequest)
 
         super.init()
+
+        if (signalMessage as? TSOutgoingMessage) != nil {
+            self.signalMessage.addObserver(self, forKeyPath: "self.messageState", options: [.new, .old], context: nil)
```
Comment (yuliaveres) [inline Token/Models/Message.swift]: don't we want to keep keypaths as constants ?

File: Token/Models/Message.swift [hunk 8]
```diff
@@ -127,5 +127,17 @@ public class Message: NSObject, NOCChatItem {
...
+        }
+    }
+
+    public override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
+        if keyPath == "messageState" {
```
Comment (yuliaveres) [inline Token/Models/Message.swift]: and here use the constants would be better, wdyt ?

File: Token/Models/Message.swift (hunk 7 above)
Comment (Elland) [inline Token/Models/Message.swift]: Trying to get this to work, nothing so far. 😭 Ideas?

**Comment by Elland** [comment]
@3lvis known issue. Whole morning looking into it (message status update).

File: Token/Models/Message.swift [hunk 9]
```diff
@@ -127,5 +127,17 @@ public class Message: NSObject, NOCChatItem {
         self.isActionable = shouldProcess && !isOutgoing && (sofaWrapper?.type == .paymentRequest)
 
         super.init()
+
+        if (signalMessage as? TSOutgoingMessage) != nil {
```
Comment (yuliaveres) [inline Token/Models/Message.swift]: also wouldn't that be better `if let signalMessage = signalMessage as TSOutgoingMessage? { ` ...

File: Token/Models/Message.swift (hunk 3 above)
Comment (yuliaveres) [inline Token/Models/Message.swift]: that makes understanding code and debugging much easier, with one return place

File: Token/Models/Message.swift (hunk 7 above)
Comment (yuliaveres) [inline Token/Models/Message.swift]: please, try `self.signalMessage.addObserver(self, forKeyPath: #keyPath(TSOutgoingMessage.messageState),` ...

**Comment by Elland** [comment]
State tracking should be fixed now. 🚀

**Pull request merged**